src/renderer.c

```diff
@@ -137,97 +175,195 @@ static int font_set_style(FT_Outline* outline, int x_translation, unsigned char
-  unsigned int byte_width = font->antialiasing == FONT_ANTIALIASING_SUBPIXEL ? 3 : 1;
-  for (int j = 0, pen_x = 0; j < bitmaps_cached; ++j) {
-    GlyphSet* set = check_alloc(calloc(1, sizeof(GlyphSet)));
-    font->sets[j][idx] = set;
-    for (int i = 0; i < GLYPHSET_SIZE; ++i) {
-      int glyph_index = FT_Get_Char_Index(font->face, i + idx * GLYPHSET_SIZE);
-      if (!glyph_index || FT_Load_Glyph(font->face, glyph_index, load_option | FT_LOAD_BITMAP_METRICS_ONLY)
-        || font_set_style(&font->face->glyph->outline, j * (64 / SUBPIXEL_BITMAPS_CACHED), font->style) || FT_Render_Glyph(font->face->glyph, render_option)) {
-        continue;
-      }
-      FT_GlyphSlot slot = font->face->glyph;
-      unsigned int glyph_width = slot->bitmap.width / byte_width;
-      if (font->antialiasing == FONT_ANTIALIASING_NONE)
-        glyph_width *= 8;
-      set->metrics[i] = (GlyphMetric){ pen_x, pen_x + glyph_width, 0, slot->bitmap.rows, true, slot->bitmap_left, slot->bitmap_top, (slot->advance.x + slot->lsb_delta - slot->rsb_delta) / 64.0f};
-      pen_x += glyph_width;
-      font->max_height = slot->bitmap.rows > font->max_height ? slot->bitmap.rows : font->max_height;
-      // In order to fix issues with monospacing; we need the unhinted xadvance; as FreeType doesn't correctly report the hinted advance for spaces on monospace fonts (like RobotoMono). See #843.
-      if (!glyph_index || FT_Load_Glyph(font->face, glyph_index, (load_option | FT_LOAD_BITMAP_METRICS_ONLY | FT_LOAD_NO_HINTING) & ~FT_LOAD_FORCE_AUTOHINT)
-        || font_set_style(&font->face->glyph->outline, j * (64 / SUBPIXEL_BITMAPS_CACHED), font->style) || FT_Render_Glyph(font->face->glyph, render_option)) {
-        continue;
-      }
-      slot = font->face->glyph;
-      set->metrics[i].xadvance = slot->advance.x / 64.0f;
+static unsigned int font_get_glyph_id(RenFont *font, unsigned int codepoint) {
+  if (codepoint > MAX_UNICODE) return 0;
+  size_t row = codepoint / CHARMAP_COL;
+  size_t col = codepoint - (row * CHARMAP_COL);
+  if (!font->charmap.rows[row]) font->charmap.rows[row] = check_alloc(calloc(sizeof(unsigned int), CHARMAP_COL));
+  if (font->charmap.rows[row][col] == 0) {
+    unsigned int glyph_id = FT_Get_Char_Index(font->face, codepoint);
+    // use -1 as a sentinel value for "glyph not available", a bit risky, but OpenType
+    // uses uint16 to store glyph IDs. In theory this cannot ever be reached
+    font->charmap.rows[row][col] = glyph_id ? glyph_id : (unsigned int) -1;
+  }
+  return font->charmap.rows[row][col] == (unsigned int) -1 ? 0 : font->charmap.rows[row][col];
+}
+
+#define FONT_IS_SUBPIXEL(F) ((F)->antialiasing == FONT_ANTIALIASING_SUBPIXEL)
+#define FONT_BITMAP_COUNT(F) ((F)->antialiasing == FONT_ANTIALIASING_SUBPIXEL ? SUBPIXEL_BITMAPS_CACHED : 1)
+
+static SDL_Surface *font_find_glyph_surface(RenFont *font, FT_GlyphSlot slot, int bitmap_idx, GlyphMetric *metric) {
```

Comment:
The function name makes it sound like it's just doing a simple search, but in reality it's also assigning the glyph to the surface (or viceversa?).
Maybe a better name could be `font_assign_glyph_surface` or even `font_assign_surface_to_glyph`, `font_bind_glyph_to_surface`.